Adding the issue for enhancement with OKT proposition (#84)
operator-framework/enhancements: enhancements/orange_okt_proposition.md

	- **Mutator** - Here I tell to OKT to apply the Mutation on all resources present in its registry (load Initial/defaults data, then apply CR values if needed). 
	- **Updater** - Here OKT, thanks to a hash algorithm on resources data, will compute if after Mutation, the resource has to be Created, Update, or is unchanged against it's cluster peer instance. So for all resource in its registry it apply the same idempotent process to update (or not!!) the resource.
	- **SuccessManager** - If all goes right (no error) , this stage is reached, and here I'll can tell to OKT to "manageSuccess()" i.e. check for the right requeuing value to return and complete this reconciliation but at same time perform the update of a status condition for the Reconciliation, transparently for me, if I passed the CR.Status when I instatiate my OKT Reconciler object. 
	- At any previous stage, if something goes wrong, an error can be raised, there's also the case where we giveup the reconciliation for whatever reason (not an error). There is also the case where a CR Finalization is triggered. These "debranching" cases are also present in my Reconciliation steps function thanks to 2 dedicated steps: 

> 		-    **CRFinalizer** - here I can perform my own tasks when OKT detected that a finalizer (with the right name) exists and the CR is being deleted.

Author: Actually, the OKT Reconciler just check if a finalizer in CR Metadata with the same name as itself exists. If it's the case and the CR is being deleted, the stage "CRFinalizer" is called right after the stage "CRChecker"  instead of continuing the normal process (ObjectsGetter, ...).
The principle is to do what you need in the CRFinalizer stage and call a provided method to delete the finalizer.
So it is just a way to manage a single finalizer on the CR. 
I guess the helper you mentioned is generic as well to manage several finalizers...

> 		-    **ErrorManager** - This stage can be reached after any other stages, and let me call the OKT "manageError()" method that will pickup the last error and return the right requeuing value to complete this reconciliation.

Author: Yes a good explanation by a sample would help. This the the reason I'll provide a more complete version of the Memcached operator sample. I believed Eric S. got a tarball with the first version of Memcached aligned with the "User Story 1" in this issue, feel free to go it. 
Now, I'm very close to get the official GO to publish OKT on Github, I think there's no issue to drop a new Memcached operator sample under my Github space.  I'll try to release it nex week.

In Story 1 we described a Reconciliation cycle triggered at each event and trying to traverse a list of steps (a branch) as follow :

    CRChecker->ObjectsCreator->Mutator->Updator->ManageSuccess  (+ 2 "debranching" steps to ManageError & CRFinalizer)
Going from 1 step to the other is conditionned by the success of all actions taken during the step. Else we debranch to the `ManageError` step. All of this happen during **1 Reconciliation cycle**.

For my application lifecycle, I have 1 graph (name it **App LC Graph**) of steps representing the applications states I want to manage. At each step some actions have to be done, that may take a while: 

     Start->Running->Servicing
                   ->Stopping
                   -> End
Going from 1 step to the other is conditionned by some conditions that may be met  **over N Reconciliation cycles**.

I like the idea to have a clear view on the steps I defined previously, so I'll complete my work with the OperatorSDK and the OKT addon.

OKT comes with a statemachine feature that should help in defining these steps and let me focus on the code I need to implement at each step. 
To allow this, OKT provides:

>   -  a sidecar for my application to help me to get my database status and launch actions on it asynchronously.

Author: I'm not sure to well capture your point but I guess you wondering if the CR status is the right place to store the [conditions](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go)  that represent the application's state. I had in mind the ability to offer  a somple way to follow the application state thanks to a "kubectl wait" command for example. But actually it sounds a bit unsuitable to me too, so may be that was the reason of your question :). 
So it's a point still unstated today.  Besides that, building a specific interface relying on the [CloudEvent](https://github.com/cloudevents/spec) api could be a better option either in term of interoperability with external world and may be for others internal needs too...

### Non-Goals

Non goals can be all that is already or should be covered by OLM.

## Proposal

### User Stories [optional]

#### Story 1

Here a first use-case simulation.

I want to implement a Kubernetes Operator in GO and it will have to manage 3 resources (ex a Configmap, a Secret and a Deployment).

I want to be aligned with best pratices, so I decide to use the OperatorSDK but I choose to use also the OKT (tell it like this) addon in order to get benefit of the enhanced Reconciliation process, a more straight forward CR status management and finalization, a result utility to trace what happens and prepare the reconciliation response (error + requeueing time).

1. I create a project with the Operator SDK command as usual
2. In my Controller, I choose an OKT Reconciler and add the Reconciliation steps function in which I can see distinctly the standard steps in which I should pass at each reconciliation event (each step below is an entry in an embracing `switch(step) { case "xxx": ... }`:
	- **CR Checker** - once the CR is picked up by OKT and validated by webhooks (if any), I can tell here if I have something to do more on my CR

> 	- **ObjectGetter** - Here I will add the code to create my 3, in memory, resources and add them to the OKT's registry of resources (still in memory). However at this stage OKT pick up the resource on the K8S Cluster so it knows if the resource in its registry has an existing peer on the Cluster (or cache).

Reviewer: Assuming you mean that the operator registers 3 types with OKT, this sounds a lot like controller-runtime's [cache](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/cache).

Author: I must admit don't know lot of things about the controller-runtime's cache implementation.

At OKT level it is an array of a generic resource type, that OKT uses at the  "Mutator" and "Updater" reconciliation steps to loop on each resource and call respectively the "getHashXXX(), MutateWithInitialData(), MutateWithCR()" (for mutation) and a specific "CreateOrUpdate()" OKT method.

	- At any previous stage, if something goes wrong, an error can be raised, there's also the case where we giveup the reconciliation for whatever reason (not an error). There is also the case where a CR Finalization is triggered. These "debranching" cases are also present in my Reconciliation steps function thanks to 2 dedicated steps: 
		-    **CRFinalizer** - here I can perform my own tasks when OKT detected that a finalizer (with the right name) exists and the CR is being deleted.
		-    **ErrorManager** - This stage can be reached after any other stages, and let me call the OKT "manageError()" method that will pickup the last error and return the right requeuing value to complete this reconciliation.
-    In the Controller folder I use the OKT code generator for my 3 resources I want to create. i.e.: 
   `okt-gen-resource xxxxx`
- In each file generated, I always find 3 parts to fill by myself and that will constitute a common/standard way to Mutate my resource:

> 	- A method called `getHashXXX() { // Put your code here  }`  - It allows me to define, thanks to an OKT Helper, wich part of my object I want to include in the hash computation

Reviewer: How is the hash's input created? Marhsaling the object itself? I've seen quite a few cases where a time or `resourceVersion` value is included in a hash/comparison, with an always-true result, so exposing such a function may be a source of confusion and bugs.

Author: Firstly for each resource, you have to define in the "getHashXXX()" method, the scope of objects/elements covered by the hash computation. OKT provides you an helper to define this scope. Internally it use an array of pointers with all the elements participating to the scope (the metadata you want, the whole Object's Spec part if you want, actually whatever you want). 
You return this "bag" of collected elements through the return statement of "getHashXXX()' method.

To perform the Hash computation, OKT use the spew library (`github.com/davecgh/go-spew/spew`) which follows pointers and prints actual values of the nested objects ensuring the hash does not change when a pointer changes. 
The returned hash can be used for object comparisons.

So if the developper don't add `resourceVersion` in the scope, there's no risk to meet the bug you mentionned. I know that some immutable fields (IP in Services, ...) and the Secret's data are things to take care about (OKT helper help on this point).
There is an example in the operator `Memcached with OKT` I sent to you and Tony (it can be shared).

Reviewer 2: So it sounds like the hash is defined by the developer and they can choose as much or as little in their hash. And I agree with Eric that if someone unknowingly adds an attribute that changes it will cause an infinite reconcilation loop. (we've done this ourselves :)

Author: > if someone unknowingly adds an attribute that change it will cause an infinite reconcilation loop
OKT comes with an integrated throttling mechanism in case of a same recurrent errors, but not sure it can help in such case...
Besides that, when the developer defines the hash in the "getHashXXX()" method, he uses an OKT helper generic or dedicated for the resource kind he is dealing with.  For example, the SecretMutationHelper object manages few things with StringData and Data fields. It ensures too that the hash computation covers only the Data field and not the StringData field.
Here again, I don't know if there's a more elegant solutions and if a such helper is sufficient to guarantee the problem of the changing attributes added unknowingly.
The counterpart of these helpers is the need to maintain them. In OKT their usage is optional and today, the only one that exists, is the SecretMutationHelper. I thought to need another one for the Service resource and its changing ClusterIP, but in practice all works correctly without special care, it seems it was a mistake to think that.

Author: Edited the last sentence regarding specifics vs generics helpers.

> 	- A method called `getHashXXX() { // Put your code here  }`  - It allows me to define, thanks to an OKT Helper, wich part of my object I want to include in the hash computation

Reviewer 2: ```suggestion
	- A method called `getHashXXX() { // Put your code here  }`  - It allows me to define, thanks to an OKT Helper, which part of my object I want to include in the hash computation

```

	- A method  called `mutateWithIntialData() { // Put your code here }` - It is here I fill my object structure with the initial data at creation time, with eventually shared parameters across all my resources (a label name, a network port, ...). I can use to useful functions to fill my GO structure with either a YAML template or another initialized GO structure or fill my object directly as I want.
	- A method called `mutateWithCR() { // Put your code here }` - Here I copy any CR relevant fields into my GO structure that become the "expected" object I want to create/update on my Cluster 

> - Once done, I can run my Operator localy or deploy it as usual  with the OperatorSDK commands, and I have an operator respectfull to the idempotency principle of the K8S reconciliations, a status and a finalization management out of the box.

Reviewer 2: ```suggestion
- Once done, I can run my Operator locally or deploy it as usual  with the Operator SDK commands, and I have an operator respectful to the idempotency principle of the K8S reconciliations, a status and a finalization management out of the box.

```

OKT aims to provide if it worth it, an optional resource helper. When it exists in the OKT library, it centralize some utilities for a specific kind of resource. For example, right now, a StatefulSetHelper is availaible and could evolve in the future. This last provides some basic methods or shortcut like  GetReadyPodsCount() or GetRunningPodsCount(). 
If in the future if I'll have lots of resources to create, OKT allow through an option in its function call,  to create no more than X resource max at a same time (another best practice).  

> A throttle mecanism is put automatically in place by OKT if the same error occurs indefinitely to requeue the error with a growing elapsed time.

Reviewer 2: ```suggestion
A throttle mechanism is put automatically in place by OKT if the same error occurs indefinitely to requeue the error with a growing elapsed time.

```

    CRChecker->ObjectsCreator->Mutator->Updator->ManageSuccess  (+ 2 "debranching" steps to ManageError & CRFinalizer)
Going from 1 step to the other is conditionned by the success of all actions taken during the step. Else we debranch to the `ManageError` step. All of this happen during **1 Reconciliation cycle**.

For my application lifecycle, I have 1 graph (name it **App LC Graph**) of steps representing the applications states I want to manage. At each step some actions have to be done, that may take a while: 

     Start->Running->Servicing
                   ->Stopping
                   -> End
Going from 1 step to the other is conditionned by some conditions that may be met  **over N Reconciliation cycles**.

I like the idea to have a clear view on the steps I defined previously, so I'll complete my work with the OperatorSDK and the OKT addon.

OKT comes with a statemachine feature that should help in defining these steps and let me focus on the code I need to implement at each step. 
To allow this, OKT provides:

  -  a sidecar for my application to help me to get my database status and launch actions on it asynchronously.

>   -  an utility to modelize my graph of appication states into my CRD

Reviewer 2: ```suggestion
  -  an utility to modelize my graph of application states into my CRD

```

  -  a GO type to implement this graph and transition rules that condition how I validate the transition from one step to another

In my CR I set the wished state (i.e. Servicing) I want to reach, while the current application state (i.e. is maintained in the CR status with a new Condition).

Once the application added to the OKT registry (like any other resource), the OKT Reconciler knows that it has to manage this  resource as follow:

  - on Start: Create() it!
  - on End: Delete() it!
  - on any other state: Update it!

> As any other resource, it put in place an idempotent mecanism and detect changes (and thus will do nothing during a Reconciliation if there's nothing new). Here what will trigger a change:

Reviewer 2: ```suggestion
As any other resource, it put in place an idempotent mechanism and detect changes (and thus will do nothing during a Reconciliation if there's nothing new). Here what will trigger a change:

```

 
 A state change (in the App LC  Graph) is handled asynchronously to not impact the Controller with a too long task. On such case (long task) 1 or more requeueing orders are left to wait for the observable change once done. 

It also maintain a Status condition in the CR that reflect the application current state and errors if any.

To sum up:

  - an application lifecycle is managed like an infrastucture resource from OKT's point of view, 
- a clear view on what is implemented in term of application lifecycle is provided thanks to the App LC Graph described by the CRD
  - Having all the operators in an organization built upon the same model should help human (or intelligent automates) operators to deal with several kind of K8S operators.

### Implementation Details/Notes/Constraints [optional]

Today, the Story 1 is fully operational and implementable with the current version of OKT. A partial implementation detail can be provided by the MemcachedOperatorSampleWithOKT repo code.

> The architecture to render story 2 implementable is not yet fully completed and we are wondering if this approach make sense or not.

Reviewer: I think the concept of a controller state machine makes a lot of sense, and one generic enough to register an arbitrary application lifecycle for an operator to execute would be powerful. I think such a library warrants an enhancement on its own, with a prototype ideally.

Author: Yes thank you, the concept is interresting and I have some ideas to render the LC graph implementation generic at some points. Note that the K8S resources reconciliation steps (CRChecker(), ObjectsGetter(), ..., SuccessManager() ) are already implemented thanks to a StateMachine library ([Qor transition](https://github.com/qor/transition)), it is a bit overrated because initialy I thought about a more complex reconciliation process, but it's done with that right now.
It is clear now, that it is better to separate in 2 distincts processes the K8S resources reconciliation and the Application's LC graph stages. 
Besides that, some ideas comes in mind. We could manage the state changes through a standard events formalizations as the one brought by the [CloudEvents](https://github.com/cloudevents/spec) specs and lib. Providing a standard way to observe and pilot the state change seems interesting.

Reviewer 2: :+1: agree that this state machine idea sounds quite interesting.

> The OKT library is a GO module that depends on the OperatorSDK, more specifically on the sigs and k8s.io modules aligned on those used by the OperatorSDK.

Reviewer: Can you clarify exactly what you mean by "depends on the OperatorSDK"? The operator-sdk repo contains a binary that scaffolds code from controller-runtime; perhaps you mean OKT depends on the latter?

Author: Yes there's no real dependancies except that OKT has no real sense without the usage of the OperatorSDK as OKT is done to be used inside an OperatorSDK's scaffolded project.
Besides that, each time the librairies used by the OperatorSDK evolves (sigs, ...), we have to align the OKT librairy as well.

> Upgrading the OperatorSDK version means upgrading OKT, it would be less impacting if OKT was integrated in the OperatorSDK as an internal tool box. 

Reviewer 2: I'm not sure it has to be in lock step version for version. I think OKT could have it's own versioning scheme and still be compatible with Operator SDK. I think if we made it an internal toolkit, that would mean we'd have to expose a public library from within the Operator SDK repo which we didn't want to do. I think OKT could have a life of its own alongside Operator SDK. I could potentially see it merged with [operator-lib](https://github.com/operator-framework/operator-lib/) or vice versa.

### Risks and Mitigations

> What are the risks of this proposal and how do we mitigate. Think broadly. For
> example, consider both security and how this will impact the larger Operator Framework
> ecosystem.
> 
> How will security be reviewed and by whom? How will UX be reviewed and by whom?
> 
> Consider including folks that also work outside your immediate sub-project.

Reviewer 2: ```suggestion
N/A
```

## Design Details

### Test Plan

> **Note:** *Section not required until targeted at a release.*
> 
> Consider the following in developing a test plan for this enhancement:
> - Will there be e2e and integration tests, in addition to unit tests?
> - How will it be tested in isolation vs with other components?
> 
> No need to outline all of the test cases, just the general strategy. Anything
> that would count as tricky in the implementation and anything particularly
> challenging to test should be called out.
> 
> All code is expected to have adequate tests (eventually with coverage
> expectations).

Reviewer 2: ```suggestion
OKT will have its own unit tests.
```

### Graduation Criteria

> **Note:** *Section not required until targeted at a release.*
> 
> Define graduation milestones.
> 
> These may be defined in terms of API maturity, or as something else. Initial proposal
> should keep this high-level with a focus on what signals will be looked at to
> determine graduation.
> 
> Consider the following in developing the graduation criteria for this
> enhancement:
> - Maturity levels - `Dev Preview`, `Tech Preview`, `GA`
> - Deprecation
> 
> Clearly define what graduation means.

Reviewer 2: ```suggestion
N/A
```

> #### Examples
> 
> These are generalized examples to consider, in addition to the aforementioned
> [maturity levels][maturity-levels].
> 
> ##### Dev Preview -> Tech Preview
> 
> - Ability to utilize the enhancement end to end
> - End user documentation, relative API stability
> - Sufficient test coverage
> - Gather feedback from users rather than just developers
> 
> ##### Tech Preview -> GA 
> 
> - More testing (upgrade, downgrade, scale)
> - Sufficient time for feedback
> - Available by default
> 
> **For non-optional features moving to GA, the graduation criteria must include
> end to end tests.**
> 
> ##### Removing a deprecated feature
> 
> - Announce deprecation and support policy of the existing feature
> - Deprecate the feature
> 
> ### Upgrade / Downgrade Strategy
> 
> If applicable, how will the component be upgraded and downgraded? Make sure this
> is in the test plan.
> 
> Consider the following in developing an upgrade/downgrade strategy for this
> enhancement:
> - What changes (in invocations, configurations, API use, etc.) is an existing
>   cluster required to make on upgrade in order to keep previous behavior?
> - What changes (in invocations, configurations, API use, etc.) is an existing
>   cluster required to make on upgrade in order to make use of the enhancement?
> 
> ### Version Skew Strategy
> 
> How will the component handle version skew with other components?
> What are the guarantees? Make sure this is in the test plan.
> 
> Consider the following in developing a version skew strategy for this
> enhancement:
> - During an upgrade, we will always have skew among components, how will this impact your work?
> - Does this enhancement involve coordinating behavior in the control plane and
>   in the kubelet? How does an n-2 kubelet without this feature available behave
>   when this feature is used?
> - Will any other components on the node change? For example, changes to CSI, CRI
>   or CNI may require updating that component before the kubelet.

Reviewer 2: We can just drop these sections for this proposal.

```suggestion
```

## Implementation History

Major milestones in the life cycle of a proposal should be tracked in `Implementation
History`.

## Drawbacks

The idea is to find the best form of an argument why this enhancement should _not_ be implemented.

## Alternatives

Similar to the `Drawbacks` section the `Alternatives` section is used to
highlight and record other possible approaches to delivering the value proposed
by an enhancement.

> ## Infrastructure Needed [optional]
> 
> Use this section if you need things from the project. Examples include a new
> subproject, repos requested, github details, and/or testing infrastructure.
> 
> Listing these here allows the community to get the process for these resources
> started right away.

Reviewer 2: ```suggestion
## Infrastructure Needed [optional]

TBD
```
